### PR TITLE
adds dioscuri solution which allows you to connect to Gemini 1.0 Pro …

### PR DESCRIPTION
**Pull request by jamessexton-io**
…and Gemini 1.5 Pro

### CODE REVIEW

File: sgtm/dioscuri/README.md [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+## Prerequisites
+
+-   Server Side Google Tag Manager
+-   Access to a Google Cloud project with billing enabled
+
+## Vertex AI API Setup
+
+1. In the Google Cloud console, on the project selector page, select a 
+   Google Cloud project.
+2. [Use this link to enable the Vertex AI API](https://pantheon.corp.google.com/apis/enableflow?apiid=aiplatform.googleapis.com&_ga=2.89950491.1404017710.1717753577-283650328.1717753577&project=gps-asa-uk-sgtm-demo-account)
```
Comment (tdsymonds) [inline sgtm/dioscuri/README.md:26]: This is a corp link, and contains the project name. I think this is the link you need:

https://console.cloud.google.com/apis/enableflow?apiid=aiplatform.googleapis.com

File: sgtm/dioscuri/README.md [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+3. Go to IAM settings and find the service account which is used to deploy sGTM.
+   Grant this service account the Vertex AI User permission.
+
+More details can be found on this [quickstart guide](https://cloud.google.com/vertex-ai/generative-ai/docs/start/quickstarts/quickstart-multimodal)
+
+## Server-Side GTM Setup
+
+1. Download the [Dioscuri variable template](./../src/gtm/dioscuri_get_gemini_response.tpl)
```
Comment (tdsymonds) [inline sgtm/dioscuri/README.md:34]: I think this path is wrong too. Isn't it just `./dioscuri_get_gemini_response.tpl`?

File: sgtm/dioscuri/README.md [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+   - i. Update the generation configuation settings if required. These can be used
+        to tweak the response length and other elements. See below for more details
+        on what each setting can be used for
+   - j. Save
+7. Use the variable in a tag which has a live trigger and you will see data returned
+
+## Example Use Case
+
+You are a retail client who wants to get a personalised message for a user when
+they log into your site. You have already got user data  in Firestore and it is
```
Comment (tdsymonds) [inline sgtm/dioscuri/README.md:71]: nit: two spaces between data and in

File: sgtm/dioscuri/README.md [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+Add some text prompts. In this example we added the following:
+
+1. You are a virtual salesperson for a retail company which sells clothes. Write 
+a greeting for {{Firestore User Data - First name}} and suggest the kinds of products 
+they might want to look at today. It is {{Returning Customer - boolean}} that she is
+a returning customer and it is {{Is High Value User - boolean}} that she is a high value 
+customers. They live in {{Firestore User Data - Address - City}}, {{Firestore User Data - 
+Address - Country}}.
+2. Here is some information about {{Firestore User Data - First name}} in a stringified 
+JSON format which you should parse: {{Custome Data}}
```
Comment (tdsymonds) [inline sgtm/dioscuri/README.md:83]: Custome -> Customer / Custom?

File: sgtm/dioscuri/README.md [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+![dioscuri_example_customer_data.png](./img/dioscuri_example_customer_data.png)
+
+When we run the tag you can see Gemini returns a personalised response which is 
+aware of his favourte category and colour:
+
+![dioscuri_example_response.png](./img/dioscuri_example_response.png)
+
+
+
+## Explanation of Settings
```
Comment (tdsymonds) [inline sgtm/dioscuri/README.md:100]: It sounds like these have been pulled from a cloud document somewhere. Might be worth linking that for more info

File: sgtm/dioscuri/dioscuri_get_gemini_response.tpl [hunk 6]
```diff
@@ -0,0 +1,653 @@
...
+ * @version 1.0.0
+ */
+const getGoogleAuth = require("getGoogleAuth");
+const JSON = require("JSON");
+const logToConsole = require("logToConsole");
+const sendHttpRequest = require("sendHttpRequest");
+
+//Function to convert string of words to array of strings
+function splitAndTrim(commaSeparatedString) {
+  commaSeparatedString = "hello, goodbye";
```
Comment (tdsymonds) [inline sgtm/dioscuri/dioscuri_get_gemini_response.tpl:443]: Should this line be here? It's overwriting the argument

File: sgtm/dioscuri/dioscuri_get_gemini_response.tpl [hunk 7]
```diff
@@ -0,0 +1,653 @@
...
+    parts.push({"fileData": {
+      "fileUri": filePrompt.fileUrl,
+      "mimeType": filePrompt.mimeType
+    }});
+  });
+}
+
+if (parts.length == 0) {
+  logToConsole("At least one prompt must be supplied");
+  //data.gtmOnFailure();
```
Comment (tdsymonds) [inline sgtm/dioscuri/dioscuri_get_gemini_response.tpl:487]: Why is this commented out. Seems reasonable to fail here.

Comment (jamessexton-io) [reply sgtm/dioscuri/dioscuri_get_gemini_response.tpl:487]: Worked out that data.gtmOnFailure() is not available in variables which is when you get a not defined error. Removed

File: sgtm/dioscuri/dioscuri_get_gemini_response.tpl (hunk 6 above)
Comment (jamessexton-io) [reply sgtm/dioscuri/dioscuri_get_gemini_response.tpl:443]: Was used for testing - removed

**Review by tdsymonds — APPROVED** [review]

**Pull request merged**